baybe/simulation/lookup.py

```diff
@@ -0,0 +1,209 @@
+              to one parameter configuration (columns names being the parameter names)
+              and must return a dataframe containing the corresponding target values
+              (column names being the target names), where each row holds the targets
+              for the corresponding parameter row.
+        impute_mode: Specifies how a missing lookup will be handled (only relevant
+            for dataframe lookups). See :func:`baybe.simulation.core.simulate_campaign`
+            for details.
+
+    Raises:
+        AssertionError: If an analytical function is used and an incorrect number of
+            targets was specified.
+    """
+    # If no lookup is provided, invent some fake values
+    if lookup is None:
+        add_fake_results(queries, targets)
+
+    # Compute the target values via a callable
+    elif isinstance(lookup, Callable):
+        _lookup_targets_from_callable(queries, [t.name for t in targets], lookup)
+
+    # Get the results via dataframe lookup (works only for exact matches)
+    # IMPROVE: Although its not too important for a simulation, this
+    #  could also be implemented for approximate matches
+    elif isinstance(lookup, pd.DataFrame):
+        _lookup_targets_from_dataframe(queries, targets, lookup, impute_mode)
+
+
+def _lookup_targets_from_callable(
+    queries: pd.DataFrame,
+    target_names: Collection[str],
+    lookup: CallableLookup,
+) -> None:
+    """Look up target values via a callable."""
+    # Evaluate the callable
+    responses = lookup(queries)
+
+    # Assert that all targets are contained in the response
+    if not (exp := set(target_names)).issubset(act := set(responses)):
+        raise ValueError(
+            f"The provided lookup callable yielded values for the labels {act} but "
+            f"required are values for {exp}."
+        )
+
+    # Insert the target values in-place
+    queries[responses.columns] = responses
+
+
+def _lookup_targets_from_dataframe(
+    queries: pd.DataFrame,
+    targets: List[Target],
+    lookup: CallableLookup,
+    impute_mode: Literal["error", "worst", "best", "mean", "random"] = "error",
+) -> None:
+    """Look up target values from a dataframe."""
+    target_names = [t.name for t in targets]
+    all_match_vals = []
+
+    # IMPROVE: speed up the matching using a join
+    for _, row in queries.iterrows():
+        ind = lookup[
+            (lookup.loc[:, row.index] == row).all(axis=1, skipna=False)
+        ].index.values
+
+        if len(ind) > 1:
+            # More than one instance of this parameter combination
+            # have been measured
+            warnings.warn(
+                f"The lookup rows with indexes {ind} seem to be "
+                f"duplicates regarding parameter values. Choosing a "
+                f"random one.",
+                UserWarning,
+            )
+            match_vals = lookup.loc[np.random.choice(ind), target_names]
```

Comment:
Not too happy with that solution. Is this a situation that we want to support? In any case, the user should already be informed about this at an earlier point, ideally in a suitable docstring that if the lookup table contains dulicates, then a random one will be chosen.

Reply:
Fully agree. In my opinion, this should be the user's responsibility, not the one of the function. We should simply throw an error instead and let to user make sure there are no duplicates. However, note that I did not touch the code in this PR. The function needs an entire refactor anyway, to replace the loops with dataframe joins. So I just added a TODO note for now and we postpone this to the moment when then functional issues are addressed in another PR. If OK, please resolve